substrate/client/chain-spec/src/lib.rs

```diff
@@ -16,19 +16,155 @@
 // You should have received a copy of the GNU General Public License
 // along with this program. If not, see <https://www.gnu.org/licenses/>.
 
-//! Substrate chain configurations.
+#![allow(rustdoc::private_intra_doc_links)]
+
+//! # Substrate chain configurations.
+//! This crate includes structs and utilities for defining configuration files (known as chain
+//! specification) for both runtime and node.
+//!
+//! The chain specification is a JSON file that describes the properties and intial state of the
+//! chain.
+//!
+//! In summary, though not restirced to, the main role of the chain spec is to provide a list of
+//! well-known boot nodes for the blockchain network and the means for initializing the genesis
+//! storage. This initialization is necessary for creating a genesis block upon which subsequent
+//! blocks are built. When the node is launched for the first time, it reads the chain spec,
+//! initializes the genesis block, and establishes connections with the boot nodes.
+//!
+//! It is divided into three main logical sections:
+//! - one section details general chain properties,
+//! - the other explicitly or indirectly defines the genesis storage, which, in turn,
+//! determines the genesis hash of the chain,
+//! - third deals with the runtime code.
+//!
+//! The chain specification consits of the following fields:
+//!
+//! | Chain spec key | descirption |
+//! |---------|---------|
+//! |`name`|The human readable name of the chain.|
+//! |`id`|The id of the chain.|
+//! |`chainType`|The chain type of this chain (refer to [`ChainType`]).|
+//! |`bootNodes`|A list of multi addresses that belong to boot nodes of the chain.|
+//! |`telemetryEndpoints`|Optional list of `multi address, verbosity` of telemetry endpoints.  The verbosity goes from `0` to `9`. With `0` being the mode with the lowest verbosity.|
+//! |`protocolId`|Optional networking protocol id that identifies the chain.|
+//! |`forkId`|Optional fork id. Should most likely be left empty. Can be used to signal a fork on the network level when two chains have the same genesis hash.|
+//! |`properties`|Custom properties. Shall be provided in the form of `key`, `value` json object.|
+//! |`consensusEngine`|Deprecated field. Should be ignored.|
+//! |`codeSubstitutes`|Optional map of `block_number` to `wasm_code`. More details in material to follow.|
+//! |||
+//! |`genesis`|Defines the initial state of the runtime. More details in material to follow.|
+//! |||
+//! |`code`|The runtime wasm code blob in hex format.|
+//!
+//!
+//! # Initial runtime state aka genesis storage
+//!
+//! The `genesis` key of the chain specification definition describes the
+//! initial state of the runtime. For example it may contain:
+//! - initial list of funded accoumts,
+//! - the sudo key,
+//! - initial authorities set for consensus, etc.
+//!
+//! The genesis state can be represented in the formats described by [`chain_spec::Genesis<G>`]
```

Comment:
Not sure what to do here. `chain_spec::Genesis<G>` is private (and I believe it should be).
But the list below just shows all the variants of this `enum` which is not desirable.

Reply:
You are describing here the format of the file based on some internal structures. This is wrong. The internal structures only exist to parse the format, but they don't dictate the format.